src/peft/optimizers/riemannian.py

```diff
@@ -0,0 +1,294 @@
+import math
+from operator import attrgetter
+from typing import Callable, Iterable, Tuple
+
+import torch
+import torch.nn as nn
+from torch.optim import Optimizer
+from transformers.utils.versions import require_version
+
+from ..peft_model import PeftModel
+
+
+def create_riemannian_optimizer(
+    model: PeftModel,
+    optimizer_cls: type[Optimizer],
+    optimizer_kwargs: dict,
+    lr_embedding: float = 1e-6,
+    reg: float = 1e-4,
+) -> Optimizer:
+    """
+    Creates a Riemmanian optimizer. Used only for LoRA. Implementation based on:
+    https://github.com/pilancilab/Riemannian_Preconditioned_LoRA. Paper reference: https://arxiv.org/pdf/2402.02347.
+
+    Args:
+        model (`torch.nn.Module`): The model to be optimized.
+        optimizer_cls (`torch.optim.Optimizer`): The optimizer class to be used.
+        optimizer_kwargs (`dict`): Additional keyword arguments to be passed to the optimizer.
+            - lr_embedding (`float`): The learning rate to be used for the embedding layer. Defaults to lr_embedding
+            - reg (`float`): Regularization parameter for Riemmanian preconditioner. Included for lora parameters only
+              and is needed for invertability guarantees
+    """
+
+    # TEST VERSION FOR ADAMW
+    if not issubclass(optimizer_cls, torch.optim.AdamW):
+        raise TypeError("TEST version only supports AdamW optimizer")
```

Comment:
Since the `optimizer_cls` argument is not actually except to raise an error, how about removing it completely?